File: fs/pkg/types/inode.go
package types

import (
	"fmt"
)

type Ino uint64

const (
	DirectBlocksCount Block = 12
	InodeSize         Byte  = 512
	InoSize           Byte  = 8
	InoNil            Ino   = 0
	InoRoot           Ino   = 1
)

type Inode struct {
	Ino                 Ino
	FileType            FileType
	Size                Byte
	LinksCount          uint16
	DirectBlocks        [DirectBlocksCount]Block
	SinglyIndirectBlock Block
	DoublyIndirectBlock Block
	TriplyIndirectBlock Block
}

type FileType uint8

const (
	FileTypeInvalid FileType = iota
	FileTypeRegular
	FileTypeDir
	FileTypeCharDev
	FileTypeBlockDev
	FileTypeFifo
	FileTypeSocket
	FileTypeSymlink
)

func (ft FileType) String() string {
	switch ft {
	case FileTypeInvalid:
		return "Invalid"
	case FileTypeRegular:
		return "Regular"
	case FileTypeDir:
		return "Dir"
	case FileTypeCharDev:
		return "CharDev"
	case FileTypeBlockDev:
		return "BlockDev"
	case FileTypeFifo:
		return "Fifo"
	case FileTypeSocket:
		return "Socket"
	case FileTypeSymlink:
		return "Symlink"
	default:
		panic(fmt.Sprintf("invalid file type: `%d`", ft))
	}
}

func (ft FileType) MarshalJSON() ([]byte, error) {
	s := ft.String()
	out := make([]byte, len(s)+2)
	out[0] = '"'
	out[len(out)-1] = '"'
	copy(out[1:], s)
	return out, nil
}

func (ft FileType) Validate() error {
	if ft <= FileTypeInvalid || ft > FileTypeSymlink {
		return fmt.Errorf(
			"validating file type `%d`: %w",
			ft,
			InvalidFileTypeErr,
		)
	}
	return nil
}

const (
	InvalidFileTypeErr ConstError = "invalid file type"
)

File: rustfs/src/inode.rs
use crate::block::*;
use crate::byte::*;

#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
pub struct Ino(u64);

impl Ino {
    pub fn new(value: u64) -> Ino {
        Ino(value)
    }
}

impl std::fmt::Display for Ino {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{0}", self.0)
    }
}

impl From<Ino> for Byte {
    fn from(ino: Ino) -> Byte {
        Byte::new(ino.0)
    }
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Inode {
    pub ino: Ino,
    pub size: Byte,
    pub direct_blocks: [Option<Block>; DIRECT_BLOCKS_COUNT],
    pub singly_indirect_block: Option<Block>,
    pub doubly_indirect_block: Option<Block>,
    pub triply_indirect_block: Option<Block>,
}

impl Inode {
    pub fn new(ino: Ino) -> Inode {
        Inode {
            ino: ino,
            size: Byte::new(0),
            direct_blocks: [
                None, None, None, None, None, None, None, None, None, None, None, None,
            ],
            singly_indirect_block: None,
            doubly_indirect_block: None,
            triply_indirect_block: None,
        }
    }

    pub fn set_direct_block(&mut self, index: BlockListIndex, block: Option<Block>) {
        index.set_direct_block(&mut self.direct_blocks, block)
    }
}
